runtime/major_gc.c

```diff
@@ -562,10 +581,164 @@ static void mark_ephe_aux (struct mark_stack *stk, intnat *work,
   }
 }
 
+
+#define Pb_size (1 << 8)
+#define Pb_min 64
+#define Pb_mask (Pb_size - 1)
+
+Caml_inline void prefetch_block(value v)
+{
+  /* Prefetch a block so that scanning it later avoids cache misses.
+     We will access at least the header, but we don't yet know how
+     many of the fields we will access - the block might be already
+     marked, not scannable, or very short. The compromise here is to
+     prefetch the header and the first few fields.
+
+     We issue two prefetches, with the second being a few words ahead
+     of the first. Most of the time, these will land in the same
+     cacheline, be coalesced by hardware, and so not cost any more
+     than a single prefetch. Two memory operations are issued only
+     when the two prefetches land in different cachelines.
+
+     In the case where the block is not already in cache, and yet is
+     already marked, not markable, or extremely short, then we waste
+     somewhere between 1/8-1/2 of a prefetch operation (in expectation,
+     depending on alignment, word size, and cache line size), which is
+     cheap enough to make this worthwhile. */
+  caml_prefetch(Hp_val(v));
+  caml_prefetch(&Field(v, 3));
+}
+
+Caml_inline uintnat rotate1(uintnat x)
+{
+  return (x << ((sizeof x)*8 - 1)) | (x >> 1);
+}
+
+Caml_noinline static intnat do_some_marking
+#ifndef CAML_INSTR
+  (intnat work)
+#else
+  (intnat work, int* slice_fields, int* slice_pointers)
+#endif
+{
+  uintnat pb_enqueued = 0, pb_dequeued = 0;
+  int darkened_anything = 0;
+  value pb[Pb_size];
+  /* These global values are cached in locals,
+     so that they can be stored in registers */
+  struct mark_stack stk = *Caml_state->mark_stack;
+  uintnat young_start = (uintnat)Val_hp(Caml_state->young_start);
+  uintnat half_young_len =
+    ((uintnat)Caml_state->young_end - (uintnat)Caml_state->young_start) >> 1;
+#define Is_block_and_not_young(v) \
+  (((intnat)rotate1((uintnat)v - young_start)) > (intnat)half_young_len)
```

Comment:
```suggestion
    ((uintnat)Caml_state->young_end - young_start) >> 1;
#define Is_block_and_not_young(v) \
  (((intnat)rotate1((uintnat)v - young_start)) >= (intnat)half_young_len)
```
The definition of `Is_young(x)` is equivalent to _a ≤ x < b_ for _a_ = `Val_hp(Caml_state->young_start)` and _b_ = `Caml_state->young_end`. Hacker's Delight gives me the formula _x ‑ a < b - a_  (unsigned comparison, but this is handled by your rotation trick). Whereas you seem to have written _x - a ≤ b - a + 1_.

`Val_hp(Caml_state->young_end)` is not young according to `Is_young`, whereas it is according to `Is_block_and_not_young`. Can you please have a look?

(I'm not the best at spotting and fixing off-by-one errors, so I might be mistaken; but in this case better safe than sorry because I think `Caml_state->young_end` is allowed to be the header of a major heap value as it can be past the end of the allocated array. Unless the alignment trick of `caml_stat_alloc_aligned` makes it impossible (I did not check) but anyway in that case better guard against bugs in waiting.)

Reply:
Thanks, fixed.